src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,804 @@
+		return NULL;
+
+	json_data = g_uri_unescape_string (metadata, NULL);
+	id = extract_runtime_info_from_json_data (json_data, &url, &error);
+
+	if (!id) {
+		g_debug ("Error getting external runtime from "
+			 "metadata: %s", error->message);
+		return NULL;
+	}
+
+	priv = gs_plugin_get_data (plugin);
+
+	full_id = g_strdup_printf ("%s:%s",
+				   gs_flatpak_get_prefix (priv->flatpak),
+				   id);
+
+	runtime = gs_plugin_cache_lookup (plugin, full_id);
+	if (runtime) {
+		g_debug ("Found cached '%s'", full_id);
+		/* Adopt runtime if needed */
+		if (g_strcmp0 (gs_app_get_management_plugin (runtime),
+			       gs_plugin_get_name (plugin)) != 0) {
+			gs_app_set_management_plugin (runtime,
+						      gs_plugin_get_name (plugin));
+		}
+
+		return runtime;
+	}
+
+	runtime = gs_app_new (full_id);
+	gs_app_set_metadata (runtime, METADATA_HEADLESS_APP,
+			     gs_app_get_id (headless_app));
+	gs_app_set_metadata (runtime, METADATA_URL, url);
+	gs_app_set_metadata (runtime, "flatpak::kind", "runtime");
+	gs_app_set_kind (runtime, AS_APP_KIND_RUNTIME);
+	gs_app_set_flatpak_name (runtime, id);
+	gs_app_set_flatpak_arch (runtime, flatpak_get_default_arch ());
+	gs_app_set_flatpak_branch (runtime, "master");
+	gs_app_set_management_plugin (runtime, gs_plugin_get_name (plugin));
+
+	gs_plugin_cache_add (plugin, full_id, runtime);
+
+	if (gs_flatpak_is_installed (priv->flatpak, runtime, NULL, NULL)) {
+		gs_app_set_state (runtime, AS_APP_STATE_INSTALLED);
+	}
+
+	return runtime;
+}
+
+gboolean
+gs_plugin_refine_app (GsPlugin *plugin,
+		      GsApp *app,
+		      GsPluginRefineFlags flags,
+		      GCancellable *cancellable,
+		      GError **error)
+{
+	GsApp *ext_runtime;
+	const char *metadata = NULL;
+	GsPluginData *priv;
+
+	/* We cache all runtimes because an external runtime may have been
+	 * adopted by the flatpak plugins */
+	if (app_is_flatpak (app) &&
+	    g_strcmp0 (gs_app_get_flatpak_kind_as_str (app), "runtime") == 0) {
+		gs_plugin_cache_add (plugin, gs_app_get_id (app), app);
+		g_debug ("Caching remote '%s'", gs_app_get_id (app));
+	}
+
+	/* only process this app if was created by this plugin */
+	if (g_strcmp0 (gs_app_get_management_plugin (app),
+		       gs_plugin_get_name (plugin)) != 0)
+		return TRUE;
+
+	g_debug ("Refining external app %s", gs_app_get_id (app));
+
+	ext_runtime = gs_plugin_get_app_external_runtime (plugin, app);
+
+	if (!ext_runtime) {
+		g_debug ("Could not understand the asset from the "
+			 "metadata in "
+			 "app %s: %s",
+			 gs_app_get_id_no_prefix (app), metadata);
+		return TRUE;
+	}
+
+	priv = gs_plugin_get_data (plugin);
+
+	if (!gs_flatpak_refine_app (priv->flatpak, app, flags, cancellable, 
+				    error)) {
+		g_debug ("Refining app %s failed!", gs_app_get_id (app));
+		return FALSE;
+	}
+
+	/* We set the state to available because we assume that we can build
+	 * the runtime */
+	if (gs_app_get_state (ext_runtime) != AS_APP_STATE_INSTALLED) {
+		gs_app_set_state (app, AS_APP_STATE_UNKNOWN);
```

Comment:
That's because state transitions are "regulated". An app cannot go from installed to available directly but can go from installed to unknown to available.
Just in case it's not clear, the reason we fake that the app is uninstalled is because it needs the ext. runtime which is not installed.